### DISCUSSION TITLE
Simple (?) tiling script needed

### DISCUSSION THREAD

**Discussion by AlphaTwentyThree** [ImageMagick/ImageMagick#5334] [Help]
Hi folks,

it's been some years since I've worked with ImageMagick and have since forgotten how to find my way around. What I want to do is quite simple I hope.
Create a 2x2 tiling (1.jpg, 2.jpg...):
- black 5% border around each tile
- each image inside the tile needs to be centered
- I want to have an option to resize all images to either the largest height or largest width beforehand as they can differ in dimensions quite a lot

Thanks for your help!

**Comment by fmw42** [comment]
See montage or script -smush (-append with padding)

**Reply by AlphaTwentyThree** [reply]
Well that's the thing, I've searched for anything that lets me center the images but -gravity south for example only works for convert.
To be honest, I don't get along with ImageMagick quite the way I'd like to and 90% of the commands I try don't work at all. Don't get me wrong, normally I'm not one to ask for help that quickly and I would normally like your approach of minimal - but at the moment I just don't have the capacity to get into it again; I only need it for this little task to automate a process.
So in fact I'd be happy if somebody could quickly write something for me to test if it produces the result I need and tweak it a little.

**Comment by snibgo** [comment]
What version of IM, on what platform?

I'm not sure what you need. I suppose each image should retain its aspect ratio. Perhaps each image should be resized to fit in a box that is the maximum width of all the images, and the maximum height of all the images. Hece some images will be enlarged, but none will be reduced.  The black border around each image should be 5% of what? After resizing each image and borders, the sizes will in general be different, so pixels will need to be added to make them tile 2x2. Should those pixels be black or what?

It would help if you could mock-up an example using Gimp or something.

**Reply by AlphaTwentyThree** [reply]
Yeah sorry, should have been a bit more specific, you're right.

- IM9 on Windows 10
- background color doesn't matter, I can change than manually as needed
- about the 5%: the image with the border should be 5% bigger than the raw image, so 1.05*w x 1.05*h. Or just the average of that as border thickness, so 0.025*(w+h).
Here are the example mock-ups:
1: Each row scaled to largest height of row - https://ibb.co/R2J0TY6 (all images centered horizontally)
2: Each column scaled to largest width of column - https://ibb.co/mrKSZrV (all images centered vertically)
3: Optimized to minimize background (bottom right stayed original) - https://ibb.co/Nmy8qKZ
raw pictures:
https://ibb.co/1nY8KzK
https://ibb.co/vQdPQ3x
https://ibb.co/C6RKs4y
https://ibb.co/T44DTfM

Thanks for helping! Feel free to ask any further questions!

**Comment by snibgo** [comment]
How about this (Windows BAT syntax). I have used three different colors so you can see where they cam from, if you want to adjust dimensions.
```
%IMG7%magick ^
  4up1.jpg 4up2.jpg 4up3.jpg 4up4.jpg ^
  -set option:WW %%[fx:max(u[0].w,max(u[1].w,max(u[2].w,u[3].w)))] ^
  -set option:HH %%[fx:max(u[0].h,max(u[1].h,max(u[2].h,u[3].h)))] ^
  -resize %%[WW]x%%[HH] ^
  -gravity Center ^
  -background Blue -extent %%[WW]x%%[HH] ^
  -bordercolor #77f -border %%[fx:max(WW,HH)*0.05] ^
  -background Black ^
  ( -clone 0-1 +smush %%[fx:WW*0.05] +repage ) ^
  ( -clone 2-3 +smush %%[fx:HH*0.05] +repage ) ^
  -delete 0-3 ^
  -append +repage ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/179843953-48fcfd0e-113e-4a0e-95af-f20454073d48.png)

**Reply by AlphaTwentyThree** [reply]
Thanks a lot for this! I will play around with it and get back to you soon.

**Reply by AlphaTwentyThree** [reply]
Sadly, the script doesn't work for me, it just hangs in there.

**Comment by snibgo** [comment]
It works fine for me, and takes about 1 second on my laptop, with the files you linked.

Are you using the command in a BAT script? With those files?

What version of IM? You said "IM9 on Windows 10" but IM doesn't have a version 9. Please show the text output from "magick -version".

**Reply by AlphaTwentyThree** [reply]
Ah sorry, didn't know how to get the correct version apparently:
`ImageMagick 7.0.11-4 Q16 x64 2021-03-20`

Using a normal bat file. This is the status the cmd remains on: https://ibb.co/QN4zBqN

**Reply by AlphaTwentyThree** [reply]
Edit: ok I thought it doesn't work - my PC actually takes 25 seconds for the task... Will get back to you with the results I got, sorry for the mess here.

**Reply by AlphaTwentyThree** [reply]
Alright, that's a start. Two things to clear up until we can tweak further:
1. I don't want a double border on the middle portion, it should be the same width as the border around the whole tiling - here's a mockup: https://ibb.co/4tX0yFY
2. How can I change this from height to width and finally to "optimized" (see mock-up)?

**Comment by snibgo** [comment]
1. That's why I used three different colours. If you don't want the black pixels, use `-smush 0` instead of `+smush %%[fx:WW*0.05]`.

2. I don't understand the question.

**Reply by AlphaTwentyThree** [reply]
`-smush 0` results in https://ibb.co/JqK4dV5. Did you look at my mock-up? Let me make that a bit clearer: https://ibb.co/hZH1G35. I've left the colors so you can see what I mean.

Let's continue until this is working. Thanks for your patience.

**Comment by snibgo** [comment]
Sorry, I said "-smush" when I should have said "+smush".
```
magick ^
  4up1.jpg 4up2.jpg 4up3.jpg 4up4.jpg ^
  +write info: ^
  -set option:WW %%[fx:max(u[0].w,max(u[1].w,max(u[2].w,u[3].w)))] ^
  -set option:HH %%[fx:max(u[0].h,max(u[1].h,max(u[2].h,u[3].h)))] ^
  -resize %%[WW]x%%[HH] ^
  +write info: ^
  -gravity Center ^
  -background Blue -extent %%[WW]x%%[HH] ^
  -bordercolor #77f -border %%[fx:max(WW,HH)*0.05] ^
  +write info: ^
  -background Black ^
  ( -clone 0-1 +smush 0 +repage ) ^
  ( -clone 2-3 +smush 0 +repage ) ^
  -delete 0-3 ^
  -append +repage ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/180441469-772fd2da-c543-4057-931d-b4fe6ae93e58.png)
In your annotation "need this to be half as thick", why do you say *half*? The dark blue pixels are added (by `-extent`) to make all four images the same width as the widest, and the same height as the tallest.

Extending to make all four images the same size is fairly easy, as I showed. If you want something else, then you need to decide on the rules.

**Reply by AlphaTwentyThree** [reply]
Ok, now I get what the script does. I don't want that. I want the script to compare the width of the pictures in **each column** and resize the smaller image to the bigger one, so something like maxw(1up1.jpg,1up3.jpg) for column 1. The width of both columns is **not** the same. The dark blue border in this case only appears on two of the four pictures then of course.
In the end, the distance between each single picture as well as to the edge of the tiling needs to be the same (`0.025*(W1+W2)`.). That is the **LIGHT BLUE** part (#77f), **NOT** the blue one - just like I marked on my mock-up...

**Reply by AlphaTwentyThree** [reply]
Wrong:
![x_wrong](https://user-images.githubusercontent.com/97170745/180481490-4635dc9f-9385-4fdc-8a15-584bb095bfe7.png)

Correct:
![x_correct](https://user-images.githubusercontent.com/97170745/180481501-3dcae2f3-684a-4c76-b3ab-de83ccecf232.png)

The correct one has an inner border that is, in fact, **half as wide** as the wrong one - just like I said. Basically this isn't even a tiling, just a background layer with four pictures on top in equidistance.

**Comment by snibgo** [comment]
You want to halve the thickness of the central light-blue "+" shape? Is that right? Zero smush would make that thickness twice the border, so we can smush by the negative of the border.
```
magick ^
  4up1.jpg 4up2.jpg 4up3.jpg 4up4.jpg ^
  -set option:WW %%[fx:max(u[0].w,max(u[1].w,max(u[2].w,u[3].w)))] ^
  -set option:HH %%[fx:max(u[0].h,max(u[1].h,max(u[2].h,u[3].h)))] ^
  -resize %%[WW]x%%[HH] ^
  -gravity Center ^
  -background Blue -extent %%[WW]x%%[HH] ^
  -set option:SMSH %%[fx:-max(WW,HH)*0.05] ^
  -bordercolor #77f -border %%[fx:max(WW,HH)*0.05] ^
  -background Black ^
  ( -clone 0-1 +smush %%[SMSH] +repage ) ^
  ( -clone 2-3 +smush %%[SMSH] +repage ) ^
  -delete 0-3 ^
  -smush %%[SMSH] +repage ^
  x.png
```
![x](https://user-images.githubusercontent.com/33812368/180514341-75fa5327-6e42-499a-ab20-fd2ed8d71a55.png)

> I want the script to compare the width of the pictures in each column and resize the smaller image to the bigger one, ...

So you need to calculate two different widths, one for each column. But you want only one height? Is that correct? So all four images become the same height, but image 1 and 3 become the same width, and 2 and 4 become the same width?

**Reply by AlphaTwentyThree** [reply]
Hey cool, that works so far. Slowly understanding how these commands work now. ;)

The heights of the rows will of course be different - the lower one should have `maxh(4up3.jpg,4up4.jpg)`, the upper one the maximum height **after** the width adjustment. Not all pictures I'll throw in there will have the same ratio, so I need this to work for either permutation of the source pictures in this example.

Here's a mock-up:
![Untitled](https://user-images.githubusercontent.com/97170745/180562837-289c51b8-892a-4e79-a634-81a317522cbe.png)

Notice how there's only a single picture with the blue background left. 4up3.jpg still has it's original size.

**Reply by AlphaTwentyThree** [reply]
Here's a mock-up for a different configuration:
![Untitled2](https://user-images.githubusercontent.com/97170745/180567353-83563831-a8ee-4f14-b8d4-d580c7ad67f7.png)
Only the first of the pictures was resized.

**Comment by snibgo** [comment]
So, we need to calculate two heights (one per row) and two widths (one per column).

Each input is then processed separately, with a resize and extent to the appropriate width and height. We would do that in a clone within parentheses. Then we add borders and smush them together as before.

Does that sound reasonable? I have shown how to do each element, so can you do that?

**Comment by snibgo** [comment]
No, my previous post cannot make your mock-up. I thought you wanted the top two image to fit in a box that was the height of the tallest (165 pixels) But both the top images have an (extended) height of about 221 pixels.

So, I'm afraid I can't see the rules you want to follow to make the required output.

**Comment by snibgo** [comment]
Here are some different possible rules.

1. The image at top-left doesn't change size.

2. Change the image at top-right to be the same height as the top-left, retaining aspect ratio.

3. Change the image at bottom-left to be the same width as the top-left, retaining aspect ratio.

4. Change the image at bottom-left to be the same width as the *new* width as top-right, and the same height as the *new* height of the bottom-left. This is the only image that needs extending.
```
magick ^
  4up2.jpg 4up1.jpg 4up3.jpg 4up4.jpg ^
  +write info: ^
  -set option:W0 %%[fx:u[0].w] ^
  -set option:H0 %%[fx:u[0].h] ^
  ( -clone 1 -resize x%%[H0] ) ^
  ( -clone 2 -resize %%[W0] ) ^
  -set option:W1 %%[fx:u[4].w] ^
  -set option:H2 %%[fx:u[5].h] ^
  -background Blue -gravity Center ^
  ( -clone 3 -resize %%[W1]x%%[H2] -extent %%[W1]x%%[H2] ) ^
  -delete 1-3 ^
  -bordercolor #77f -border %%[fx:max(W1,H2)*0.05] ^
  ( -clone 0-1 +smush 0 +repage ) ^
  ( -clone 2-3 +smush 0 +repage ) ^
  -delete 0-3 ^
  -smush 0 +repage ^
  x.png
```

![x](https://user-images.githubusercontent.com/33812368/180590429-c14090fd-37bc-4979-8433-c75a2a7b18a7.png)
